exercises/forth/example.go

```diff
@@ -0,0 +1,243 @@
+// Package forth implements a tiny subset of the Forth language.
+package forth
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+type operatorFn func(stack *[]int) error
+
+type operatorId byte
+
+const (
+	addOp operatorId = iota
+	subOp
+	mulOp
+	divOp
+	dropOp
+	dupOp
+	swapOp
+	overOp
+	constOp
+	userDefOp
+	endDefOp
+)
+
+type operatorTyp struct {
+	fn operatorFn
+	id operatorId
+}
+
+func Forth(input []string) (result []int, err error) {
+	if len(input) == 0 {
+		return result, nil
+	}
+
+	stack := make([]int, 0, 8)
+	userDefs := make(map[string][]operatorTyp, 8)
+	for _, phrase := range input {
+		// Parse one phrase of input, building up an operator list.
+		opList, err := parse(phrase, userDefs)
+		if err != nil {
+			return nil, err
+		}
+		// Perform any operators from that phrase, updating stack.
+		for _, opr := range opList {
+			err = opr.fn(&stack)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return stack[:len(stack)], nil
+}
+
+// parse given phrase, returning an operator list, and updating
+// a userDefs map for any user definifion of words in phrase.
+func parse(phrase string, userDefs map[string][]operatorTyp) (oplist []operatorTyp, err error) {
+	words := strings.FieldsFunc(phrase,
+		func(r rune) bool {
+			return unicode.IsSpace(r) || unicode.IsControl(r)
+		})
+
+	var t int // t is token index into words[]
+	for t = 0; t < len(words); t++ {
+		w := strings.ToUpper(words[t])
+		// Handle reference to user defined word.
+		if udef, ok := userDefs[w]; ok {
+			oplist = append(oplist, udef...)
+		} else if op, ok := builtinOps[w]; ok {
+			if op.id == userDefOp {
+				// Handle user defined word definition.
+				t++
+				if t >= len(words)-2 {
+					return nil, emptyUserDefErr
+				}
+				userword := strings.ToUpper(words[t])
+				_, numerr := strconv.Atoi(userword)
```

Comment:
personal style, so optional: I consider `if _, numerr := strconv.Atoi(userword); numerr == nil` , to make it clear that `numerr` is never used after this, otherwise my brain has to keep track of whether it might be used afterward.